BUGFIX: fix relative imports

Caldwell-74 opened the pull request:
Snarling added relative pathing a while back 
but currently for imports it only works partially

the only part that works is in the editor when hovered over the import 
![image](https://github.com/bitburner-official/bitburner-src/assets/115591472/e5180d20-cab1-42ff-8f29-f009fc01ab8a)

neither the ram calculation nor the compiler uses the relative path

this pr/draw attempts to fix this

caveats:
i havent done much testing with this only used the testscripts ill provide at the end
and variable names are temporary

i just want to get feedback first because this touches sensible areas of the codebase

```js
// test/test2.js
import { test3 } from "test1.js"
/** @param {NS} ns */
export async function main(ns) {
  ns.tprint(test3())
}
// test/test1.js
export function test3(){
  return "test"
}
```

Commit:
first draw

WIP but working

Snarling commented on src/Script/RamCalculations.ts:
```diff
@@ -338,7 +344,7 @@ function parseOnlyCalculateDeps(code: string, currentModule: string): ParseDepsR
       {
         ImportDeclaration: (node: Node, st: State) => {
-          const importModuleName = node.source.value;
+          const importModuleName = resolveFilePath(node.source.value, filename) as FilePath;
```
Instead of asserting that this is a FilePath instead of null, why not throw if it ends up being null so we can control the error message?

Just based on eyeballing the code, reasons this might end up resolving as null would be player used an invalid character in an import filename, or player used too many ../ and tried going up one level from the root directory, etc. Asserting a real FilePath here if it's possibly null is going to lead to a weird error somewhere else due to us lying to TS

Caldwell-74 replied:
oh yeah that was just to get it to work
ill it cleanup today

Caldwell-74 committed:
fixed file extensionless resolving and changed path resolving

ScriptEditorContextShape gets a FilePath so we need a little code duplication to make sure that the conversion can be handled if it fails
changed resolveFilePath to resolveScriptFilePath

Commit:
fix relative imports in ramCalculations

d0sboots commented:
My high-level feedback: You're doing good work. If you can get relative paths to work in a sensible, consistent fashion, that would be wonderful. 

However, there is a higher bar than usual for touching this fiddly part of the code: add at least one new test (or test case to an existing test) that fails without your change and passes with it. In other words, a test to validate proper functioning of relative imports, now and into the future.

Caldwell-74 commented:
Thanks.

I will definetly try myself in writing tests. Its just an area i havent ventured in yet so i put it on the end.

Current hold up is the magic string "__SPECIAL_INITIAL_MODULE__"
i need confirmation that this isnt used by another package/the engine 
and that it can safely be replaced by the main scripts file name
because if so it will simplify the code and i wont have to do ugly stuff like 
`parseCode(script.code, nextModule, nextModule);`
CatLover tested it briefly and it does seem to work so i can change that back again and then start doing the test

d0sboots commented:
I would `grep -r` to be sure, but I'm pretty sure it's only used here (because that's the point).

Caldwell-74 committed:
clean up and error message improvements

Caldwell-74 committed:
fix broken function calls in tests

Caldwell-74 committed:
StaticRamParsing tests

Added 3 new tests
1. One layer deep relative importing
2. Two layer deep relative importing
3. Two layer deep relative importing but in a different folder with a potential resolving conflict

Caldwell-74 committed:
npm run doc

Caldwell-74 committed:
more consistent naming

d0sboots commented on src/Script/RamCalculations.ts:
```diff
@@ -92,7 +101,7 @@ function parseOnlyRamCalculate(otherScripts: Map<ScriptFilePath, Script>, code:
   }
 
   // Parse the initial module, which is the "main" script that is being run
-  const initialModule = "__SPECIAL_INITIAL_MODULE__";
```
There's some comment references to this that aren't fixed up.

Caldwell-74 replied:
changed it to <name of the main script> gonna push it when the other stuff is resolved

d0sboots commented on src/Script/RamCalculations.ts:
```diff
@@ -101,21 +110,19 @@ function parseOnlyRamCalculate(otherScripts: Map<ScriptFilePath, Script>, code:
 
   // Finally, walk the reference map and generate a ram cost. The initial set of keys to scan
-  // are those that start with __SPECIAL_INITIAL_MODULE__.
+  // are those that start with the name of the main script.
```
What happens when the initial script is `/foo.js` and there is `/foo.js.bar.js`?

Caldwell-74 replied:
```js
import { test } from "./foo.js.bar"
/** @param {NS} ns */
export async function main(ns) {
  ns.tprint(test(ns))
}
//foo.js.bar.js
export function test(ns) {
  return ns.getServer("home")
}
```

tried with this and it seems to just work?
i dont fully understand that lower section of the code where initialModule is used so if there is a special case you see that needs testing other than that let me know

d0sboots replied:
Yeah, I was legitimately asking because I don't know either. I'm vaguely concerned about this case, but I may have to look into it more.

d0sboots commented on src/ScriptEditor/ui/ScriptEditorContext.tsx:
```diff
@@ -28,14 +31,20 @@ export function ScriptEditorContextProvider({ children, vim }: { children: React
     }
-
-    const ramUsage = calculateRamUsage(newCode, server.scripts);
+    //temp duplication to turn FilePath into ScriptFilePath
```
Explain this comment, please?

Caldwell-74 replied:
the editor gets scripts and non script files (like txt)
for those something isnt included like no newCode
we cant resolve scriptFilePath before we have checked that its a script 
but after the resolving it its possible that the script is null (shouldnt happen but ts needs a check)
so i need to duplicate the same error handling like right above it because i havent found a better way yet
i planned to just make a dirty fix there and come back to it when iam (mostly/fully) done with the relative path stuff 
and then make a proper

d0sboots replied:
I think what you're looking for is to add `hasScriptExtension` to the checks above. As a type assertion, you will then be able to use the ContentFilePath as a ScrptFilePath.

d0sboots commented on src/ScriptEditor/ui/ScriptEditorRoot.tsx:
```diff
@@ -143,6 +143,7 @@ function Root(props: IProps): React.ReactElement {
     infLoop(newCode);
     updateRAM(
       !currentScript || currentScript.isTxt ? null : newCode,
+      currentScript?.path ?? null,
```
`?? null` should be unneeded here?

Caldwell-74 replied:
`if (newCode === null || filename === null || server === null) {`
it explicity looks for null in updateRAM thats why i assumed it has to be null not undefined
and i tried to keep the changes (for now/this pr) in that section to as little as possible

d0sboots replied:
But currentScript can't be undefined? 

If you're worried about it, switch to `== null` to catch both nullish values.

Caldwell-74 replied:
its intialized with null thats why the other two arguments check for truthyness of currentscript there
its actually a hazard because we have like a dozen of null/truthyness checks for it because of that

i would like to change that when as a whole but assumed thats out of scope of this pr

edit.: 
oh the undefined would come from the optional chaining operator not from openscript itself

d0sboots replied:
Ah, I thought optional chaining passed the value through (i.e. `(null)?.foo === null`), but I see that was wrong. But regardless, you can handle it by changing the check in the function to `== null` (either for just this term or all of them),  which catches null and undefined both.

Caldwell-74 replied:
i would go with 
`currentScript && currentScript.path,`
when thats okay cuz then it fits style wise to the other check
i can also change the === null's to == null

d0sboots replied:
`currentScript && currentScript.path` is the same as `currentScript?.path` but less clear/more bug prone IMO. For instance it would be 0 if currentScript were 0; that shouldn't happen, but that's why nullish coalescing exists.

Edit: The other terms have other considerations that make nullish coalescing inappropriate for them (like not wanting to pass undefined to GetServer). I don't think it's important for them to all look the same, but if you want that then OK.

d0sboots commented on test/jest/Netscript/StaticRamParsingCalculation.test.ts:
```diff
@@ -315,8 +344,119 @@ describe("Parsing NetScript code to work out static RAM costs", function () {
+        `;
+      const libScriptTwo = new Script(libNameTwo, libCodeTwo);
+
+      const fail_libCodeTwo = `
```
I wouldn't call this ( and variables like it) "fail", because nothing about this file actually fails, does it?

Caldwell-74 replied:
the bug i try to test here for was something i completly overlooked at first 
(resolving imports on the second/n layer based of the path of the main script and not the importing scripts path)
called it "fail" because the test would fail when it loads that file by mistake kinda a trap file

d0sboots replied:
I would call it 'incorrect" or "notLoaded" instead of "fail", then. Tests should always be written from the perspective of assuming they pass :D (But with appropriate logging to help when they fail)

Caldwell-74 replied:
but iam not that happy with the naming in the tests anyway i can take a second look on that tomorrow or if you have any names/naming scheme you like to see iam happy to implement that

d0sboots reviewed (commented):
Looking pretty good!

Commit:
partial review fix

Caldwell-74 committed:
more review fixes

d0sboots commented:
I think the only thing left here is that I want to do some more detailed testing of my own when I can get back to my computer.

d0sboots commented:
OK, finally managed to do the testing I wanted. Everything seems to working as expected. Bare imports (`import { foo } from "barename.js"`) resolve to the root directory, which is correct/necessary to prevent breakage.
